py/gc.c

```diff
@@ -477,29 +477,21 @@ static void gc_deal_with_stack_overflow(void) {
     }
 }
 
-static void gc_sweep(void) {
-    #if MICROPY_PY_GC_COLLECT_RETVAL
-    MP_STATE_MEM(gc_collected) = 0;
-    #endif
-    // free unmarked heads and their tails
-    int free_tail = 0;
-    #if MICROPY_GC_SPLIT_HEAP_AUTO
-    mp_state_mem_area_t *prev_area = NULL;
-    #endif
+// Run finalisers for all to-be-freed blocks first
+// (necessary during gc_sweep_all() to avoid accessing already-freed memory)
+static void sweep_run_finalisers(void) {
```

Comment:
I'd call this `gc_sweep_run_finalisers()`.

Also, I think it's necessary to run the finalisers fist for all GC sweeps, not just `gc_sweep_all`: if object X has a finaliser and is not reachable, then all of the things Y it points to are also not reachable.  If Y is before X in the GC heap then it'll be reclaimed before `X.__del__` is run.

This has probably never been an issue in practice because objects with finalisers don't access Y.

(And the corresponding commit comment would need to be updated to reflect this situation, if what I say above is true.)

Reply:
You're right, it's currently easy to reproduce the failure with `gc_sweep_all()` but it could happen with other object graphs.